File: perceptron.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "perceptron.h"

static int calc_check(int *x_vec, int *beta_vec, int *target) {
	return *target * (beta_vec[0] + beta_vec[1] * x_vec[0] + beta_vec[2] * x_vec[1]);
}

static void update_hyperplane(int *x_vec, int *beta_vec, int *target) {
	size_t beta_vec_size = sizeof(beta_vec) / sizeof(beta_vec[0]) + 1;
	printf("%s: beta_vec_size: %zu\n", __func__, beta_vec_size);

	beta_vec[0] = beta_vec[0] + ALPHA * *target;
	for (size_t i = 1; i < beta_vec_size; i++) {
		beta_vec[i] = beta_vec[i] + ALPHA * *target * x_vec[i-1];
	}
}

int main(int argc, char **argv) {
	int data_set[MATRIX_ROWS][MATRIX_COLS];
	int beta_vec[1][MATRIX_COLS];
	const int data_set_size = sizeof(data_set) / sizeof(data_set[0]);
	const int beta_vec_size = sizeof(beta_vec) / sizeof(beta_vec[0]);
	time_t time;
	data_set_entry *data_ptr = data_set;

	// TEST!
	parseData(data_set, "x_data", "data.txt");
	parseData(beta_vec, "beta_data", "data.txt");

	int errors = 0;
	int steps_performed = 0;
	int num_updates = 0;
	time = clock(); // start time
	do {
		errors = 0;
		for (int i = 0; i < beta_vec_size; i++) {
			printf("[%d][*] beta_vec: (%d, %d, %d)\n", steps_performed, beta_vec[i][0],
					beta_vec[i][1], beta_vec[i][2]);
		}
		for (int i = 0; i < data_set_size; i++) {
			int *x_vec = data_ptr[i];
			int target = data_ptr[i][MATRIX_COLS-1];

			printf("[%d][%d] data_set_vec: (%d, %d, %d)\n", steps_performed, i, data_ptr[i][0],
					data_ptr[i][1], target);

			int result = calc_check(data_ptr[i], beta_vec[0], &target);
#ifdef VERBOSE
			printf("[%d][%d] result: %d\n", steps_performed, i, result);
#endif

			if (result <= 0) {
				errors++;
				num_updates++;
				printf("[%d][%d] errors++, updating the hyperplane!\n", steps_performed, i);
				update_hyperplane(x_vec, beta_vec[0], &target);
			}
		}
		steps_performed++;
		// safety check
		if (steps_performed > 20) {
			printf("20 steps performed, bailing out...\n");
			break;
		}
	} while (errors > 0);

	time = clock() - time; // end time
	double time_taken = ((double)time) / CLOCKS_PER_SEC; // in seconds
	time_taken *= 1000; // millisecs

	printf("FINAL HYPERPLANE: %d + %d * x1 + %d * x2\n", beta_vec[0][0], beta_vec[0][1], beta_vec[0][2]);
	printf("TOTAL num of updates: %d\n", num_updates);
	printf("TOTAL execution time: %f msecs\n", time_taken);
}

File: perceptron.py
"""
[0] -- x1
[1] -- x2
[2] -- target val (aka class)
"""
data_set   = [[1, 1, 1],
              [3, 1, 1],
              [1, 3, -1],
              [3, 3, -1]]

data_set2   = [[1, 1, 1],
               [1, -1, -1],
               [-1, -1, 1],
               [-1, 1, -1]]

# init defaults
ALPHA       = 1
beta0       = -1
beta_entry1 = 1
beta_entry2 = -1


def calc_check(classifier, x1_val, x2_val):
    return classifier * (beta0 + beta_entry1*x1_val + beta_entry2*x2_val)


# Select for 2nd exercise
#data_set = data_set2

errors = 1
total_updates = 0
steps_performed = 0;
while errors > 0:
    steps_performed += 1
    errors = 0
    for i in range(0, len(data_set)):
        x1_val = data_set[i][0]
        x2_val = data_set[i][1]
        y_val  = data_set[i][2]

        print("x_val: (" + str(x1_val) + ", " + str(x2_val) + ")")
        print("y_val: " + str(y_val))

        result = calc_check(y_val, x1_val, x2_val)
        if result <= 0:
            errors += 1
            total_updates += 1
            print("result of calc_check is <= 0: " + str(result))
            print("errors++, need to update the hyperplane!")
            beta0 = beta0 + ALPHA * y_val
            beta_entry1 = beta_entry1 + ALPHA * y_val * x1_val
            beta_entry2 = beta_entry2 + ALPHA * y_val * x2_val
            print("new beta0: " + str(beta0))
            print("new beta: (" + str(beta_entry1) + ", " + str(beta_entry2) + ")")

print("FINAL HYPERPLANE: " + str(beta0) + " + " + str(beta_entry1) + " * x1 + " + str(beta_entry2) + " * x2")
print("NUM OF UPDATES: " + str(total_updates));

